padinfo/boardgenerator.py

```diff
@@ -0,0 +1,69 @@
+class BoardGenerator(object):
+    """Generate a Dawnglare link with optional inversion."""
+
+    allowed_letters = "RGBLDHXJMPZ "
+    inversion = "-1"
+    dawnglare_link = "https://pad.dawnglare.com/?patt={}&showfill=1"
+
+    def __init__(self, board):
+        self.board = board.split()
+        self.board_height = None
+        self.board_width = None
+        self.invalid_board = False
+        self.invert = BoardGenerator.inversion in self.board
+        self.clean_board = self.get_clean_board()
+        self.invalid_orbs = not all(letter in BoardGenerator.allowed_letters for letter in self.clean_board)
+        self.board_length = len(self.clean_board)
+        self.eval_board_size()
+        self.link = self.get_link()
+
+    def get_clean_board(self):
+        board = self.board
+
+        if self.invert:
+            board.remove(BoardGenerator.inversion)
+            clean = ''.join(self.invert_fill(board))
+        else:
+            clean = ''.join(board)
+
+        return clean
+
+    def invert_fill(self, board):
+        inverted_board = []
+
+        try:
+            for i in range(len(board[0])):
+                build_string = ""
+                for j in range(len(board)):
+                    build_string += board[j][i]
+                inverted_board.append(build_string)
+        except:
+            self.invalid_board = True
+
+        return inverted_board
+
+    def eval_board_size(self):
+        length = self.board_length
+
+        if length == 20:
+            height, width = 4, 5
+        elif length == 42:
+            height, width = 6, 7
+        elif length == 30:
+            height, width = 5, 6
+        else:
+            height = 0
+            width = 0
+            self.invalid_board = True
+
+        self.board_height = height
+        self.board_width = width
```

Comment:
You don't have to set local variables before defining the class constants, it's ok to immediately:

`self.board_height, self.board_width = 4, 5`

Reply:
I was doing that to make the declarations shorter for functions where I was using that variable a lot. But if that's against best practices, I can change it.

Reply:
ah, I'd say that's superseded by "avoid unnecessary local variables." This is a bit of a judgment call. Elsewhere you aliased `board = self.board` and I didn't say "don't do that" because I know you did that to make it easier to read, and probably in 6 months you won't find it necessary to do that, but for now, seeing long "variable" names like `self.board` can actually make the code harder to read and `board = self.board` is arguably necessary.

If you have something like `data["player"]["stats"]["item"]` and you're referencing it 3 times in the same function, to get `data["player"]["stats"]["item"]["id"]` and `data["player"]["stats"]["item"][is_null]` and `data["player"]["stats"]["item"]["name"]` then I will 100% alias this as `item = data["player"]["stats"]["item"]` at the start because that's way too hard to read.

so this is a tradeoff between "how long is an annoyingly long parameter to read vs avoid local variable names"

Having only `self.param_name` vs `param_name` is generally considered *not* so long that you have to alias it, and especially in this case, the presence of your local variable names leads the reader of your code to think that you will be using the local variables rather than just assigning them at the end of the function and then exiting